### PR TITLE
PoC arbitrary dispatch

### PR DESCRIPTION
**Pull request by pmeier**
@NicolasHug @vfdev-5 following our offline discussion. This is just a PoC for `F.equalize` / `transforms.RandomEqualize`. I have to clean up a little, but the general structure is there.

### CODE REVIEW

File: torchvision/transforms/v2/_transform.py [hunk 1]
```diff
@@ -1,27 +1,32 @@
...
 from torchvision.utils import _log_api_usage_once
 
 
 class Transform(nn.Module):
-
-    # Class attribute defining transformed types. Other types are passed-through without any transformation
-    # We support both Types and callables that are able to do further checks on the type of the input.
-    _transformed_types: Tuple[Union[Type, Callable[[Any], bool]], ...] = (torch.Tensor, PIL.Image.Image)
+    # This needs to be populated by the subclasses
+    _USED_DISPATCHERS: Set[Type] = {}
```
Comment (NicolasHug) [inline torchvision/transforms/v2/_transform.py:18]: Not sure about the pros and cons of making this `None` instead... Are we cool with mutable class attributes? Maybe we can think of that later

(If the answer is that this makes mypy unhappy then hmmmmmmmm)

File: torchvision/transforms/v2/_transform.py [hunk 2]
```diff
@@ -1,27 +1,32 @@
...
 from torchvision.transforms.v2.utils import check_type, has_any, is_simple_tensor
 from torchvision.utils import _log_api_usage_once
 
 
 class Transform(nn.Module):
-
-    # Class attribute defining transformed types. Other types are passed-through without any transformation
-    # We support both Types and callables that are able to do further checks on the type of the input.
-    _transformed_types: Tuple[Union[Type, Callable[[Any], bool]], ...] = (torch.Tensor, PIL.Image.Image)
+    # This needs to be populated by the subclasses
```
Comment (NicolasHug) [inline torchvision/transforms/v2/_transform.py]: we don't want to populate that of the base class right?

```suggestion
    # This needs to be re-defined by the subclasses
```

File: torchvision/transforms/v2/functional/_color.py [hunk 3]
```diff
@@ -557,6 +557,25 @@ def autocontrast(inpt: datapoints._InputTypeJIT) -> datapoints._InputTypeJIT:
...
 
 
+def equalize(inpt: datapoints._InputTypeJIT) -> datapoints._InputTypeJIT:
+    if not torch.jit.is_scripting():
+        _log_api_usage_once(equalize)
+
+    if torch.jit.is_scripting() or is_simple_tensor(inpt):
+        return equalize_image_tensor(inpt)
+    elif isinstance(inpt, PIL.Image.Image):
+        return equalize_image_pil(inpt)
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_color.py:567]: We can worry about all that later, but will we be able to also register kernels for simple tensors and for pil? Or are we going to hit JIT issues?

File: torchvision/transforms/v2/_color.py [hunk 4]
```diff
@@ -289,6 +289,8 @@ class RandomEqualize(_RandomApplyTransform):
 
     _v1_transform_cls = _transforms.RandomEqualize
 
+    _USED_DISPATCHERS = {F.equalize}
```
Comment (NicolasHug) [inline torchvision/transforms/v2/_color.py:292]: QQs:

1. Do we expect more than one dispatcher per transform?
2. does this need really to be a set?

File: torchvision/transforms/v2/_transform.py [hunk 5]
```diff
@@ -1,27 +1,32 @@
...
 
     def __init__(self) -> None:
         super().__init__()
         _log_api_usage_once(self)
 
+        dispatchers = self._USED_DISPATCHERS.copy()
+        types = _get_supported_datapoint_classes(dispatchers.pop())
+        for dispatcher in dispatchers:
+            types &= _get_supported_datapoint_classes(dispatcher)
+        self._supported_datapoint_types = types
```
Comment (NicolasHug) [inline torchvision/transforms/v2/_transform.py:28]: Is this just

```py
types = {}
for dispatcher in dispatchers:
    types.update(_get_supported_datapoint_classes(dispatcher))
```
?


and `_get_supported_datapoint_classes` probably doesn't need to return a set if we do that

**Review by NicolasHug — COMMENTED** [review]
Thanks Philip, looks good, I'm just not sure about the need for sets

**Comment by pmeier** [comment]
Most of our transforms don't touch `_transformed_types` at all. Here is a list of the ones that do

- Dispatcher does not support arbitrary dispatch and would error for unsupported datapoints. Thus, the transform has to filter them out
  - `RandomErasing`
  - `FiveCrop`
  - `TenCrop`
  - `UniformTemporalSubsample`
  - `Normalize`
  - `RandomPhotometricDistort` (This is only partially the right category. The transform uses many dispatchers, but also [defines a "dispatcher" itself](https://github.com/pytorch/vision/blob/3966f9558bfc8443fc4fe16538b33805dd42812d/torchvision/transforms/v2/_color.py#L244-L257) that doesn't support arbitrary dispatch. If pytorch/vision#7624 goes through, this transform needs to go into the "performance" category below)
- The dispatcher supports arbitrary dispatch, but the transform filters for performance reasons (not sure the second half is true, but I can't think of anything else)
  - `Grayscale`
  - `RandomGrayscale`
- Transform has no corresponding dispatcher and thus has to filter itself. We should look into why this is the case. Maybe it has a good reason, but maybe it is just an oversight?
  - `LinearTransformation`
- Transform only applies to a single type by design. The corresponding dispatcher has no kernels registered
  - `ConvertBoundingBoxFormat`
  - `ClampBoundingBox`
- Transform converts the actual type and thus applies only to a few specific types by design. The corresponding dispatcher has no kernels registered
  - `ToTensor`
  - `PILToTensor`
  - `ToImageTensor`
  - `ToImagePIL`
- Transform supports tensor images, but no PIL images:
  - `ConvertImageDtype`
  - `ToDtype`
  - `LinearTransformation` (we could implement support here similar to what we do with `GaussianBlur`: [convert to tensor, apply the tensor kernel, and convert back](https://github.com/pytorch/vision/blob/3966f9558bfc8443fc4fe16538b33805dd42812d/torchvision/transforms/v2/functional/_misc.py#L155-L160))
  - `Normalize`

Apart from these groups, there is also `Lambda`. It takes the supported types as parameter. Nothing would change with the design in the PR, since it already circumvents `_transformed_types`.

File: torchvision/transforms/v2/_transform.py (hunk 1 above)
Comment (pmeier) [reply torchvision/transforms/v2/_transform.py:18]: No, we can go with either.

File: torchvision/transforms/v2/functional/_color.py (hunk 3 above)
Comment (pmeier) [reply torchvision/transforms/v2/functional/_color.py:567]: I would love that, since we would remove the special casing. I don't think JIT will an issue as long as we have a 

```python
if torch.jit.is_scripting():
    return ...
```

at the top.

File: torchvision/transforms/v2/_color.py (hunk 4 above)
Comment (pmeier) [reply torchvision/transforms/v2/_color.py:292]: > Do we expect more than one dispatcher per transform?

Yes. For example

- https://github.com/pytorch/vision/blob/3966f9558bfc8443fc4fe16538b33805dd42812d/torchvision/transforms/v2/_color.py#L162-L177
- https://github.com/pytorch/vision/blob/3966f9558bfc8443fc4fe16538b33805dd42812d/torchvision/transforms/v2/_geometry.py#L919-L927

> does this need really to be a set?

No. What would you prefer?

File: torchvision/transforms/v2/_transform.py (hunk 5 above)
Comment (pmeier) [reply torchvision/transforms/v2/_transform.py:28]: I wanted to make sure that we only go for the common ground. Meaning, we only transfer types that are support by _all_ dispatchers. This is why I have the intersection `&` in there.

Comment (NicolasHug) [reply torchvision/transforms/v2/_transform.py:28]: It makes we wonder about the semantic and the use of `_supported_datapoint_types`.. what if there's a transform with multiple dispatchers where each of them handle a different part of the input? Would this transform end-up being a passthrough for everything?

Comment (NicolasHug) [reply torchvision/transforms/v2/_transform.py:28]: I mean typically if we're using the intersection, we can't just have a simple logic in the transforms like

```py
if type(inpt) not in self._supported_datapoint_types:
    # pass-through inpt
```

We can't use the union either because that would pass some inputs to some dispatchers that don't support them (and error).

Seems like we have to implement the pass-through logic on a per-dispatcher basis instead (within the transforms).

Comment (pmeier) [reply torchvision/transforms/v2/_transform.py:28]: > Seems like we have to implement the pass-through logic on a per-dispatcher basis instead (within the transforms).

You are right. Let me see how this fits in
